### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by FRossouw**
Added new features:
Search functionality
Fixed issues

### CODE REVIEW

File: src/app/artists/artists.component.ts [hunk 1]
```diff
@@ -9,14 +9,32 @@ import { Artist } from '../models/artist';
 })
 export class ArtistsComponent implements OnInit {
 
-  artists: Artist[];
+  searchInput: string;
+  artists: Artist[] = new Array();
```
Comment (strizzwald) [inline src/app/artists/artists.component.ts:13]: Please try to use one variable for fetching artists. Looks like `artists` and  `artists_list` are storing the same response.

File: src/app/artists/artists.component.ts [hunk 2]
```diff
@@ -9,14 +9,32 @@ import { Artist } from '../models/artist';
...
   artists_list: Artist[] = new Array();
   constructor(private artistService: ArtistService) { }
 
   ngOnInit(): void {
-    this.artistService.getArtists(2, 22)
+    this.defaultArtists(2, 22);
+  }
+
+  searchArtist(): void {
```
Comment (strizzwald) [inline src/app/artists/artists.component.ts:21]: We could simplify here. Consider this approach:

``` typescript
search$: BehaviorSubject = new BehaviorSubject('');

updateSearch() {
    // Anytime this method is called (onKeyup), we take the value from the input
    // and put it in an observable stream, which we will use to make our HTTP requests
    // to the Deezer API.
    this.search$.next(this.searchInput);
}

ngOnInit(): void {
   
   this.search$.pipe(
        startWith(''), // We will start with an empty string here to fetch our initial artists in range 2, 22 on 1st page load.
        debounceTime(300),
        // Here, if the user enters text, we pass the text to the searchArtists(query: string) method.
        // If not, we call our "default" method getArtists(2, 22)
        map(x => (!x || !x.trim()) ? this.artistsService.getArtists(2, 22) : this.artistsService.searchAritsts(x),
        // SwitchAll is like mergeMap, except that it throws away previously emitted values from the stream
        switchAll())
        .subscribe(artists => this.artists = artists)
}
```

Then in your HTML, you could update to:

```html
<input type="text" [(ngModel)]="searchInput" (keyup)="updateSearch()" placeholder="Search for an artist" />
```
Here's the [switchAll](https://rxjs-dev.firebaseapp.com/api/operators/switchAll) docs.

File: src/app/models/deezer-error.ts [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export class DeezerError {
+    error: {
```
Comment (strizzwald) [inline src/app/models/deezer-error.ts:2]: Do we need the `error` property here? I was thinking we could maybe flatten this to:

``` typescript
export class DeezerError {
    type: string;
    message: string;
    code: number;
}
```

File: src/app/models/search.ts [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { Artist } from './artist';
+
+export class Search {
```
Comment (strizzwald) [inline src/app/models/search.ts:3]: Looks like this object is just wrapping an array of artists. Consider just using `Aritst[]` in its place.

File: src/app/services/artist.service.ts [hunk 5]
```diff
@@ -22,4 +23,11 @@ export class ArtistService {
         toArray());
   }
 
+  searchArtist(name: string): Observable<Search> {
+    const data = this.http.get<Search>(`${environment.apiSearch}${name}"`)
```
Comment (strizzwald) [inline src/app/services/artist.service.ts:27]: This `Search` model can be simplified by just replacing it with `Artist[]`

File: src/app/artists/artists.component.ts (hunk 1 above)
Comment (FRossouw) [reply src/app/artists/artists.component.ts:13]: Removed artists_list variable

File: src/app/services/artist.service.ts (hunk 5 above)
Comment (FRossouw) [reply src/app/services/artist.service.ts:27]: The results would have a data property that needs to be accessed, thus a model for this was created which stores an array of artists within a data property.

File: src/app/models/search.ts (hunk 4 above)
Comment (FRossouw) [reply src/app/models/search.ts:3]: This class holds a property that is necessary for the resulting data from the API call.
It stores an array of artists within this model object.

File: src/app/artists/artists.component.ts (hunk 2 above)
Comment (FRossouw) [reply src/app/artists/artists.component.ts:21]: There are various errors that are propagated with this method. I will continue to look into it going ahead. At this point, I will finish the rest of the errors for the review.

`
export class ArtistsComponent implements OnInit {

  searchInput: string;
  artists: Artist[] = new Array();
  search$: any;
  constructor(private artistService: ArtistService) { }

  ngOnInit(): void {
    this.search$.pipe(
      startWith(''),
      debounceTime(300),
      map(x => (!x || !String(x).trim()) ? this.artistService.getArtists(2, 22) : this.artistService.searchArtist(String(x))),
      switchAll()
    ).pipe()
      .subscribe(artists => this.artists = artists);
  }

  updateSearch() {
    this.search$.next(this.searchInput);
  }

}
`

**Pull request closed without merging**